caffe2::StorageImpl use at::DataPtr

cpuhrsch opened the pull request:
Summary: See title

Differential Revision: D9658503

Test canaries
https://our.internmc.facebook.com/intern/experiment_store/experiment/13065316/#commit1-commit2
https://our.internmc.facebook.com/intern/experiment_store/experiment/13061592/#commit1-commit2

ezyang commented on caffe2/core/storage.h:
```diff
@@ -55,12 +55,12 @@ class CAFFE2_API StorageImpl : public c10::intrusive_ptr_target {
       data_ptr_.reset(src, d);
     }
-    capacity_ = capacity;
+    numel_ = capacity / data_type_.itemsize();
```
Test it's divisible?

ezyang commented on caffe2/core/storage.h:
```diff
@@ -140,13 +140,14 @@ class CAFFE2_API StorageImpl : public c10::intrusive_ptr_target {
       data_ptr_.reset(src, d);
     }
-    capacity_ = capacity;
+    assert(data_type_.id() == data_type.id());
```
This is not the idiomatic way to assert in the Caffe2 codebase; they use CAFFE_ENFORCE instead.

ezyang commented on caffe2/contrib/aten/aten_op_template.h:
```diff
@@ -94,11 +94,13 @@ class ATenOp : public Operator<Context> {
-        });
+        at::DataPtr(
+            (void*)src.data_ptr(),
+            static_cast<void*>(&src),
```
This looks wrong. If you do this, you're going to attempt to `delete` a pointer to a stack allocated variable. That's very bad juju. You *need* to do a dynamic allocation here, if you want to make this work.

ezyang commented on caffe2/python/pybind_state_dlpack.h:
```diff
@@ -110,14 +110,16 @@ class DLPackWrapper {
-        ((int8_t*)dlTensor->data) + dlTensor->byte_offset,
+        at::DataPtr(
+            (void*)(((int8_t*)dlTensor->data) + dlTensor->byte_offset),
+            static_cast<void*>(&dlMTensor),
```
Don't take the address here. Also, the type is wrong: what you are passing here is a `DLManagedTensor**` but in the deleter you cast it to `DLManagedTensor*`

ezyang commented on caffe2/core/storage.h:
```diff
@@ -77,117 +75,150 @@ class CAFFE2_API StorageImpl : public c10::intrusive_ptr_target {
+      //         " is not divisible by itemsize " +
+      //         caffe2::to_string(data_type_.itemsize()) + " of type name " +
+      //         data_type_.name() + ".");
+      numel_ = capacity / data_type_.itemsize();
```
It might be (marginally) safer to round up rather than down here, because in that case you will be guaranteed to provide enough bytes to store whatever was previously requested.

ezyang replied:
(This is of course assuming that you don't have the `CAFFE_ENFORCE`. I'm not saying you have to do this, but if you suspect there's some off by one error on the boundary, I'd definitely make this change.)

ezyang commented on caffe2/core/storage.h:
```diff
@@ -77,117 +75,150 @@ class CAFFE2_API StorageImpl : public c10::intrusive_ptr_target {
-      data_ptr_.reset(src, d);
-    }
-    capacity_ = capacity;
+    data_ptr_ = at::DataPtr(src, data_ptr_.device());
```
Err.... you probably shouldn't be discarding the `MemoryDeleter` entirely here...

ezyang commented on caffe2/core/tensor.h:
```diff
@@ -509,17 +509,48 @@ class CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
+    if (!capacity) {
+      capacity = numel_ * data_type.itemsize();
+    }
+    at::DataPtr data_ptr = at::DataPtr(src, nullptr, d, GetDeviceType());
```
The deleter is intended to be called on `src`, so you need to set src as both the data and the context pointer, e.g., `src, src, d, GetDeviceType()`

ezyang commented on caffe2/core/tensor.h:
```diff
@@ -608,27 +638,36 @@ class CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
-            });
+        ptr = ptr_and_deleter.first;
+        deleter = ptr_and_deleter.second;
+        storage_.set_data_ptr(at::InefficientStdFunctionContext::makeDataPtr(
```
It might be worth making a context specifically for this use-case. It will save you an extra indirection, assuming that `std::function` isn't storing the closed over variables inline (unlikely), and will also you save you from any `std::function` performance stupidity.

ezyang commented on caffe2/core/tensor.h:
```diff
@@ -608,27 +638,36 @@ class CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
         auto dtor = storage_.dtype().dtor();
+        void* ptr;
+        at::DeleterFnPtr deleter;
         auto ptr_and_deleter =
```
Can we easily make this function return a `DataPtr`? That will help out the code below a bit.

ezyang commented on caffe2/contrib/aten/aten_op_template.h:
```diff
@@ -89,16 +89,18 @@ class ATenOp : public Operator<Context> {
   void assignTo(Tensor* dst, const at::Tensor& src_) {
-    at::Tensor src = src_.contiguous();
-    auto at_sizes = src.sizes();
+    at::Tensor* src = new at::Tensor(src_.contiguous());
```
I am not 100% sure what exactly you want to do here, but if it's a `Tensor` you want as context, you should directly use `TensorImpl*`. `unsafeReleaseTensorImpl`, and then `c10::raw::intrusive_ptr::decref()` inside the deleter.

cpuhrsch replied:
I just want to make sure I'm changing the code in such a way that at::DataPtr is constructed appropriately.

ezyang commented on caffe2/contrib/aten/aten_op_template.h:
```diff
@@ -89,16 +89,18 @@ class ATenOp : public Operator<Context> {
-          return src.reset();
-        });
+        at::DataPtr(
+            (void*)src->data_ptr(),
```
You don't need a cast here, right?

ezyang replied:
(Also, prefer a C++ style cast `static_cast` over a C-style cast `(foo)`)

ezyang commented on caffe2/contrib/aten/aten_op_template.h:
```diff
@@ -89,16 +89,18 @@ class ATenOp : public Operator<Context> {
+            [](void* t_ptr) -> void { delete static_cast<at::Tensor*>(t_ptr); },
+            src->device()),
+        typeMetaFor(*src),
+        0);
```
You... didn't make things any worse, but I find it hard to believe this ever worked with `capacity = 0`

cpuhrsch replied:
I think this function might actually be unused aside from assignListStartingAt which is also unused - checking right now.

ezyang commented on caffe2/core/storage.h:
```diff
@@ -16,52 +16,55 @@
+    if (numel > 0) {
+      CAFFE_ENFORCE(
+          data_type.id() != TypeIdentifier::uninitialized(),
+          "Constructing a storage with meta of unknown type and non-zero numel");
```
nit: The old error message seemed better; it told you how to fix the problem, as opposed to this one, which just tells you there's a problem

ezyang commented on caffe2/core/storage.h:
```diff
@@ -77,117 +80,135 @@ class CAFFE2_API StorageImpl : public c10::intrusive_ptr_target {
+  void set_dtype(const TypeMeta& data_type) {
+    int64_t capacity = numel_ * data_type_.itemsize();
     data_type_ = data_type;
+    numel_ = capacity / data_type_.itemsize();
```
I think there are still problems here. We should not have this method at all.

Consider the following sequence of events:
1. I allocate a tensor of `std::string`
2. On the storage, I call `set_dtype` to `int`

If I have a tensor of `std::string`, I **must** placement-delete the entries before I finish the conversion to, otherwise I will leak the data in question. So a raw call to `set_dtype` must have arranged for placement-delete to have happened; it's an INVARIANT that must be obeyed by the caller, otherwise you will leak.

ezyang commented on caffe2/core/storage.h:
```diff
@@ -77,117 +80,135 @@ class CAFFE2_API StorageImpl : public c10::intrusive_ptr_target {
-    }
-    capacity_ = capacity;
+    data_ptr_ = std::move(data_ptr);
+    // NOTE: data_type might change and so it's also possible that capacity
```
I think this comment is out of date. It is true that `data_type` may change, but in this case, we're simultaneously changing `data_type` while we're swapping out the data pointer, so we don't care about the old capacity; we're just setting the capacity based on what the user told us is the new capacity (in bytes!!) It might not be divisible by itemsize(), but then that's because the user gave us a weirdly rounded region of memory.

cpuhrsch committed:
caffe2::StorageImpl use at::DataPtr (#11282)

Summary:
See title
Pull Request resolved: https://github.com/pytorch/pytorch/pull/11282

Reviewed By: ezyang

Differential Revision: D9658503

Pulled By: cpuhrsch

fbshipit-source-id: 0bacd88c95266e7cb7e77488d1ba2f104a0c9f55
